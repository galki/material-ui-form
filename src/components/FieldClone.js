// @flow

import React from 'react'
import _ from 'lodash'


function getRequiredProp(
  required: ?boolean,
  useNativeRequiredValidator: boolean
): boolean {
  if (!useNativeRequiredValidator) {
    return false
  }
  return required || false
}

function makeLabel(fieldComp: Object, props: Object): string {
  const label: string = fieldComp.props.label || ''
  return props.field.isRequired && !props.useNativeRequiredValidator
    ? `${label} *`
    : label
}

function makeErrorAndHelperText(props: Object): Object {
  let helperText: ?string = _.get(props.fieldComp.props, 'helperText')
  let isError: boolean = false

  if (!_.isEmpty(props.field) && props.field.validations.length > 0) {
    helperText = props.field.validations[0].message
    isError = true
  }
  return { helperText, isError }
}

type Props = {
  field?: Object,
  fieldComp: Object,
  onConstruct: Function,
  onValueChange: Function,
  useNativeRequiredValidator: boolean,
  validateInputOnBlur: boolean,
};

type State = {
  helperText: ?string,
  isError: boolean,
  value: mixed,
};

export default class FieldClone extends React.Component<Props, State> {
  static defaultProps = {
    field: {},
  }

  constructor(props: Object) {
    super(props)
    const { fieldComp } = props

    if (fieldComp.type.name === undefined) {
      throw new Error('FieldClone does not support native elements')
    }
    if (fieldComp.props.name === undefined || fieldComp.props.value === undefined) {
      throw new Error('FieldClone name and value must be defined')
    }

    const value = _.isEmpty(props.field) ? fieldComp.props.value : props.field.value
    const { helperText, isError } = makeErrorAndHelperText(props)

    this.state = {
      helperText,
      isError,
      value,
    }

    if (props.field.value === undefined) {
      this.props.onConstruct(fieldComp.props)
    }
  }

  componentWillReceiveProps(nextProps: Object) {
    if (!_.isEmpty(nextProps.field)) {
      const { helperText, isError } = makeErrorAndHelperText(nextProps)
      this.setState({
        helperText,
        isError,
        value: nextProps.field.value,
      })
    }
  }

  onBlur = (event: SyntheticInputEvent<Element>) => {
    const {
      fieldComp,
      fieldComp: { props: { name } },
      validateInputOnBlur,
    } = this.props
    const { value } = event.target
    // // /* TODO: create function for condition */
    if (validateInputOnBlur && !fieldComp.props.select) {
      this.props.onValueChange(name, value)
    }
    if (fieldComp.props.onBlur !== undefined) {
      fieldComp.props.onBlur(value, { name }, event)
    }
  }

  onChange = (event: SyntheticInputEvent<Element>) => {
    const {
      fieldComp,
      fieldComp: { props: { name } },
      validateInputOnBlur,
    } = this.props
    const { value } = event.target
    if (fieldComp.props.select || validateInputOnBlur) {
      const helperText: ?string = _.get(fieldComp.props, 'helperText')
      this.setState({ isError: false, helperText, value })
    }
    /* TODO: create function for condition */
    if (!validateInputOnBlur || fieldComp.props.select) {
      this.props.onValueChange(name, value)
    }
    if (fieldComp.props.onChange !== undefined) {
      fieldComp.props.onChange(value, { name }, event)
    }
  }

  render() {
    const { fieldComp, ...props } = this.props
    return React.cloneElement(fieldComp, {
      value: this.state.value,
      label: makeLabel(fieldComp, props),
      error: this.state.isError,
      helperText: this.state.helperText,
      onBlur: this.onBlur,
      onChange: this.onChange,
      required: getRequiredProp(
        fieldComp.props.required,
        this.props.useNativeRequiredValidator
      ),
    })
  }
}
