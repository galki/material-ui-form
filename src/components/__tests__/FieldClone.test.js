import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import TextField from 'material-ui/TextField'

import FieldClone from '../FieldClone'


describe('<FieldClone>:<TextField>', () => {
  const field = {
    isPristine: true,
    isRequired: null,
    pristineValue: null,
    validations: [],
    validators: [],
    value: undefined,
  }
  const onFieldConstruct = jest.fn()
  const onFieldToggle = jest.fn()
  const onFieldValueChange = jest.fn()

  const wrapper = shallow(
    <FieldClone
      field={field}
      onConstruct={onFieldConstruct}
      onToggle={onFieldToggle}
      onValueChange={onFieldValueChange}
      useNativeRequiredValidator={false}
    >
      <TextField
        label="Name"
        type="text"
        name="name"
        value="mufasa"
        data-validators="isRequired,isAlpha"
      />
    </FieldClone>
  )

  it('should render', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should have a single child', () => {
    expect(wrapper.find(TextField)).toHaveLength(1)
  })

  it('should not throw if field type prop is undefined', () => {
    function checkTypeName() {
      if (wrapper.instance().props.children.type.name === undefined) {
        throw new Error('FieldClone does not support native elements')
      }
    }
    expect(checkTypeName).not.toThrow()
  })

  it('should not throw if field component name and value are defined', () => {
    function testNameAndValueProps() {
      const { name, value } = wrapper.instance().props.children.props
      if (name === undefined || value === undefined) {
        throw new Error('FieldClone name and value must be defined')
      }
    }
    expect(testNameAndValueProps).not.toThrow()
  })

  it('should set state', () => {
    expect(wrapper.state()).toMatchObject({
      checked: null,
      helperText: undefined,
      isError: false,
      value: undefined,
    })
  })

  it('should call onConstruct', () => {
    expect(wrapper.instance().props.onConstruct).toHaveBeenCalled()
  })

  it('should have a rendered child with an undefined value prop', () => {
    expect(wrapper.prop('value')).toBeUndefined()
  })

  it('should have a rendered label', () => {
    expect(wrapper.state('checked')).toBeNull()
    expect(wrapper.prop('label')).toBeDefined()
  })

  it('should handle onChange events', () => {
    const value = 'x'
    const event = { target: { value } }
    wrapper.find(TextField).simulate('change', event)
    expect(wrapper.state()).toMatchObject({
      helperText: undefined,
      isError: false,
      value,
    })
  })

  it('should handle onBlur events', () => {
    const value = 'x'
    const event = { target: { value } }
    const { onValueChange } = wrapper.instance().props
    wrapper.find(TextField).simulate('blur', event)
    expect(onValueChange).toBeCalledWith(wrapper.prop('name'), value)
  })

  it('should update props', () => {
    const value = 'x'
    wrapper.setProps({ field: { value, validations: [] } })
    expect(wrapper.instance().props.field.validations).toBeDefined()
    expect(wrapper.state('value')).toEqual(value)
  })
})
