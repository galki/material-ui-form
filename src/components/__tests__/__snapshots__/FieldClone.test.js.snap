// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldClone>:<TextField> should render 1`] = `
<TextField
  data-validators="isRequired,isAlpha"
  error={false}
  label="Name"
  name="name"
  onBlur={[Function]}
  onChange={[Function]}
  required={false}
  select={false}
  type="text"
/>
`;
